massdriver-application-helm/main.tf

```diff
@@ -1,5 +1,11 @@
 locals {
   base_helm_additional_values = {
+    commonLabels = module.application.params.md_metadata.default_tags
+    pod = {
+      annotations = {
+        "md-deployment-id" = lookup(module.application.params.md_metadata, "deployment_id", "")
```

Comment:
@coryodaniel FYI

We need this for 2 reasons:
* Annotations trigger a recreation of pods (don't update in place), so this effectively triggers Kubernetes own CI/CD
* This also triggers Terraform to see that the helm values have changes, which will cause it to run an `update`. Otherwise, users might update helm templates, but terraform will have no idea anything has changed since it doesn't store a hash of the templates in its state.  This ensures we get a fresh execution of helm every time somebody clicks "Deploy"

Reply:
So even if the deployment id doesn't change, if this helm values change this will pick that up?

Reply:
> So even if the deployment id doesn't change, if the* helm values change this will pick that up?

I'm not sure I follow, but perhaps I can explain better.

If the helm chart changes (the actual template files), but the *helm values* do not, terraform will *not* see that as a change and therefore will not execute and Update.  It will be a no-op.  Thats because terraform doesn't store the helm templates in the state file, only the values. And terraform will only execute an update if the "desired state" differs from the "remote state".  All terraform sees in terms of "remote state" is the last set of helm values that were used during a run.

This deployment ID is a unique UUID that gets generated every time someone clicks "Save and Deploy", which essentially means *every* deploy will look like it requires an update now.

Technically we could (perhaps should??) generate a hash of the template files as the annotation instead and this would be better since if neither the templates nor the values change, its effectively a no-op.  In our situation it will still trigger a cycling of the pods (due to a new value in the pod annotations).

Reply:
Following up, this will actually work:

```
> sha1(join("", [for f in fileset(path.cwd, "chart/*"): filesha1(f)]))
"68cbbbfca448f16e5d0e7ecacf00fb19447cc077"
```

This works *except* if the user is using `latest` as the image tag (which is an anti-pattern, but is still used quite often) since it will still look like a no-op, even if there is a new image available.  The `deployment-id` will solve even the `latest` situation since I think kubernetes requires `ImagePullPolicy` to be `always` if the image tag is `latest`, so restarting the pod (which an annotation of `deployment-id` will *always* do) will actually pull the newest image.